File: rabbit_feeder.py
#!/usr/bin/env python2
'''Listens for changes on a rethinkdb table and sends them over a
RabbitMQ topic exchange.

The topic will be <tablename>.<type_of_change> where type_of_change is
create, delete or update.'''

import pika
import rethinkdb as r
import json

# Setup RethinkDB connection
rethink_conn = r.connect(host='localhost', port=28015)

# Ensure correct database and table exist for this example
try:
    r.db_create('change_example').run(rethink_conn)
except r.RqlRuntimeError:
    pass

try:
    r.db('change_example').table_create('mytable').run(rethink_conn)
except r.RqlRuntimeError:
    pass

# Setup rabbit connection and exchange
rabbit_conn = pika.BlockingConnection(
    pika.ConnectionParameters(host='localhost', port=5672)
)
exchange = 'rethinkdb'
channel = rabbit_conn.channel()
channel.exchange_declare(exchange, exchange_type='topic', durable=False)
    

def type_of_change(change):
    '''Determines whether the change is a create, delete or update'''
    if change['old_val'] is None:
        return 'create'
    elif change['new_val'] is None:
        return 'delete'
    else:
        return 'update'

# Start feeding...
table_changes = r.db('change_example').table('mytable').changes()

for change in table_changes.run(rethink_conn):
    routing_key = 'mytable.' + type_of_change(change)
    print 'RethinkDB -(', routing_key, ')-> RabbitMQ'
    channel.basic_publish(exchange, routing_key, json.dumps(change))

File: rabbit_feeder.js
#!/usr/bin/env node

r = require('rethinkdb');
amqp = require('amqplib');

var rethinkConn = null;
var rabbitConn = null;
var channel = null;
var exchange = 'rethinkdb';


r.connect({host: 'localhost', port: 28015}).then(function(conn) {
    // Setup RethinkDB connection
    rethinkConn = conn;
}).finally(function createDB(){
    // Ensure database exists for this example
    return r.dbCreate('change_example').run(rethinkConn);
}).finally(function createTable(){
    // Ensure table exists for this example
    return r.db('change_example').tableCreate('mytable').run(rethinkConn)
}
// ignore db/table exists error
).catch(r.Error.RqlRuntimeError, function(){}
).then(function(){
    // Setup rabbit connection
    return amqp.connect('amqp://localhost:5672');
}).then(function(conn){
    rabbitConn = conn;
    // Setup rabbit channel
    return rabbitConn.createChannel();
}).then(function(ch){
    channel = ch;
    // Setup rabbit exchange
    return channel.assertExchange(exchange, 'topic', {durable: false});
}).then(function(){
    // Listen for changes on our table
    return r.db('change_example').table('mytable').changes().run(rethinkConn);
}).then(function(changeCursor){
    // Feed changes into rabbit
    changeCursor.each(function(error, change){
        var routingKey = 'mytable.' + typeOfChange(change);
        console.log('RethinkDB -(', routingKey, ')-> RabbitMQ')
        channel.publish(
            exchange, routingKey, new Buffer(JSON.stringify(change)));
    })
})

function typeOfChange(change) {
    // Determines whether the change is a create, delete or update
    if(change.old_val === null){
        return 'create';
    } else if(change.new_val === null){
        return 'delete';
    } else {
        return 'update';
    }
    return 'something'
}
